arangod/Cluster/FollowerInfo.cpp

```diff
@@ -219,147 +153,299 @@ Result FollowerInfo::remove(ServerID const& sid) {
+    if (_followers->size() + 1 < _docColl->minReplicationFactor()) {
+      _canWrite = false;
+    }
+    // we are finished
+    LOG_TOPIC("be0cb", DEBUG, Logger::CLUSTER)
+        << "Removing follower " << sid << " from " << _docColl->name() << "succeeded";
+    return agencyRes;
+  }
+  if (agencyRes.is(TRI_ERROR_CLUSTER_NOT_LEADER)) {
+    // Next run in Maintenance will fix this.
+    return agencyRes;
+  }
+
+  // rollback:
+  _followers = oldFollowers;
+  _failoverCandidates = oldFailovers;
+  std::string errorMessage =
+      "unable to remove follower from agency, timeout in agency CAS operation "
+      "for key " +
+      _docColl->vocbase().name() + "/" + std::to_string(_docColl->planId()) +
+      ": " + TRI_errno_string(agencyRes.errorNumber());
+  LOG_TOPIC("a0dcc", ERR, Logger::CLUSTER) << errorMessage;
+  agencyRes.resetErrorMessage<std::string>(std::move(errorMessage));
+  return agencyRes;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief clear follower list, no changes in agency necessary
+//////////////////////////////////////////////////////////////////////////////
+
+void FollowerInfo::clear() {
+  WRITE_LOCKER(canWriteLocker, _canWriteLock);
+  WRITE_LOCKER(writeLocker, _dataLock);
+  _followers = std::make_shared<std::vector<ServerID>>();
+  _failoverCandidates = std::make_shared<std::vector<ServerID>>();
+  _canWrite = false;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief check whether the given server is a follower
+//////////////////////////////////////////////////////////////////////////////
+
+bool FollowerInfo::contains(ServerID const& sid) const {
+  READ_LOCKER(readLocker, _dataLock);
+  auto const& f = *_followers;
+  return std::find(f.begin(), f.end(), sid) != f.end();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief Take over leadership for this shard.
+///        Also inject information of a insync followers that we knew about
+///        before a failover to this server has happened
+////////////////////////////////////////////////////////////////////////////////
+
+void FollowerInfo::takeOverLeadership(std::vector<std::string> const& previousInsyncFollowers) {
+  // This function copies over the information taken from the last CURRENT into a local vector.
+  // Where we remove the old leader and ourself from the list of followers
+  WRITE_LOCKER(canWriteLocker, _canWriteLock);
+  WRITE_LOCKER(writeLocker, _dataLock);
+  // Reset local structures, if we take over leadership we do not know anything!
+  _followers = std::make_shared<std::vector<ServerID>>();
+  _failoverCandidates = std::make_shared<std::vector<ServerID>>();
+  // We disallow writes until the first write.
+  _canWrite = false;
+  // Take over leadership
+  _theLeader = "";
+  _theLeaderTouched = true;
+  TRI_ASSERT(_failoverCandidates != nullptr && _failoverCandidates->empty());
```

Comment:
```suggestion
  TRI_ASSERT(_failoverCandidates->empty());
```
make_shared will never return a nullptr